경북대 BE 이원준 2단계 - 관리자 화면

lwj0831 opened the pull request:
제가 시험일정으로 1단계 리뷰를 늦게 요청드려 피드백을 받은지 얼마 안되어 아직 1단계 피드백을 반영하지 못했습니다..
2단계 리뷰 요청이 밀릴까봐 PR올립니다.
다음 리뷰 요청 전까지 리팩토링하겠습니다!

lwj0831 committed:
docs: README.md 작성

lwj0831 committed:
feat: domain 정의

lwj0831 committed:
chore: domain 변경

Product, Category, ProductImage의 연관관계 구현 추후 ORM도입 시 릭팩토링 예정

lwj0831 committed:
feat: 공통 API응답을 위한 Response 정의

lwj0831 committed:
feat: 상품 단건 조회, 상품 전체 조회 구현

lwj0831 committed:
feat: 상품 등록 api 개발

lwj0831 committed:
feat: 상품 수정 api 개발

lwj0831 committed:
feat: 상품 삭제 api 개발

lwj0831 committed:
feat: GlobalExceptionHandler 정의

lwj0831 committed:
refactor: 상품 조회 api Pagination 추가

lwj0831 committed:
fix: 상품 추가 api 오류 수정

id가 null로 저장되는 bug fix

lwj0831 committed:
refactor: GetProductResDto class->Record로 변경 및 생성자 대신 정적
팩터리 메서드로 변경

-복잡한 로직 없이 단순 dto역할이라 Record로 변경
-메서드 명이 제한적인 생성자 대신 정적 팩터리 메서드로 메서드 역할
분명하게 표현하기 위해 변경

lwj0831 committed:
chore: UpdateProductReqDto에 @NotBlank 추가

lwj0831 committed:
refactor: 예외 발생 부분 Repository->Service layer로 이동

예외 처리는 비즈니스 로직을 담당하는 Service에서 발생시키는 것이 바람직하다고 판단

lwj0831 committed:
refactor: InvalidSortFieldException 추가

lwj0831 committed:
Merge pull request #1 from lwj0831/refactor/product

Refactor/product

lwj0831 committed:
refactor: 상태코드를 위해 ProductController에 ResponseEntity 추가

lwj0831 committed:
refactor(build): thymeleaf 의존성 추가

lwj0831 committed:
refactor(product): GetProductResDto에 price 추가

lwj0831 committed:
refactor: application.properties -> application.yml로 변경 및 thymeleaf 설정 추가

lwj0831 committed:
feat(admin): AdminController내 상품 목록 조회 API개발 및 상품 목록 폼 개발

lwj0831 committed:
fead(admin): 상품 수정 API 및 수정 폼 개발

lwj0831 committed:
feat(admin): 상품 삭제 API 개발

lwj0831 committed:
style: README.md 2단계 내용 추가

lwj0831 committed:
Merge branch 'lwj0831' into lwj0831

lwj0831 committed:
fix: Confilct 해결 시 잘못 입력한 값 수정

liquidjoo commented on README.md:
```diff
@@ -25,6 +25,65 @@ Spring Boot 기반의 간단한 상품 관리 API입니다.
 
 ### 🧱 향후 확장 고려사항
-- [ ] 상품 검색 기능 (name 또는 카테고리 기준)
+<<<<<<< lwj0831
```
코드 병합 충돌 시 이런 부분도 확인하면 좋을 것 같아요 :)

lwj0831 replied:
병합 충돌 해결 시 좀 더 꼼꼼하게 확인해보겠습니다!!

liquidjoo commented on README.md:
```diff
@@ -25,6 +25,65 @@ Spring Boot 기반의 간단한 상품 관리 API입니다.
 - [ ] 상품 품절 상태 관리
 - [ ] 등록일 기준 정렬 기능
+
+# 🛒 2단계 - 상품 관리 웹 애플리케이션
```
요구사항 정리 👍

liquidjoo commented on src/main/java/gift/admin/controller/ProductAdminController.java:
```diff
@@ -0,0 +1,57 @@
+@Controller
+@RequiredArgsConstructor
+@RequestMapping("/admin/products")
+public class ProductAdminController {
```
<img width="1313" alt="스크린샷 2025-06-26 오후 7 13 27" src="https://github.com/user-attachments/assets/939c2a2f-7d8b-4734-88ec-102e67f3d0a7" />

<img width="576" alt="스크린샷 2025-06-26 오후 7 13 37" src="https://github.com/user-attachments/assets/5fff0f83-0ed0-4b4a-9af9-38b0bda8c758" />
<img width="681" alt="스크린샷 2025-06-26 오후 7 13 32" src="https://github.com/user-attachments/assets/cb4f4549-cf15-4fd4-834e-9e67890180ae" />

수정, 삭제가 정상적으로 작동하지 않는 것 같아요

lwj0831 replied:
Validation으로 에러 발생 시 적절한 응답할 수 있게 리팩토링하였습니다.

liquidjoo commented on src/main/java/gift/admin/controller/ProductAdminController.java:
```diff
@@ -0,0 +1,57 @@
+        return "redirect:/admin/products";
+    }
+
+    @GetMapping("/edit/{id}")
+    public String showEditForm(@PathVariable Long id, Model model) {
+        var product = productService.getProductById(id);
+        model.addAttribute("product", product);
+        return "product-edit";
+    }
+
+    @PostMapping("/edit/{id}")
+    public String updateProduct(@PathVariable Long id, @ModelAttribute UpdateProductReqDto dto) {
+        productService.updateProduct(id, dto);
+        return "redirect:/admin/products";
+    }
+
+    @PostMapping("/delete/{id}")
+    public String deleteProduct(@PathVariable Long id) {
+        productService.deleteProduct(id);
+        return "redirect:/admin/products";
+    }
```
코드 완성 후 테스트를 해보시면 좋을 것 같습니다

lwj0831 replied:
Product 도메인 클래스와 ProductRepository에 대한 unit test 추가하였습니다
이후 CI를 이용한 테스트 자동화 파이프라인 구축까지 할 계획입니다!

liquidjoo reviewed (changes requested):
안녕하세요 원준님
2단계 구현 잘해주셨어요
몇 가지 코멘트 남겼습니다 확인하고 다시 요청해 주세요 🙇

lwj0831 committed:
feat(admin): Validation 적용 안되던 문제 해결

lwj0831 committed:
chore(product): @Positive -> @PositiveOrZero로 변경

가격 0이상으로 도메인 조건 변경

lwj0831 committed:
style: Google java style guide 적용

lwj0831 committed:
chore: README.me 불필요한 부분 삭제

lwj0831 committed:
refactor: 불필요한 도메인 클래스 제거

lwj0831 committed:
refactor(product): Product 도메인 클래스 레코드로 변경 및 메서드 리팩토링

- Product 도메인 클래스의 불변성을 보장하기 위해 class에서 Record로 변경
- 기본적인 필드별 검증 로직 추가
- 정적 팩터리 메서드 사용으로 메서드명을 통해 명확한 메서드 동작 파악 및
  검증로직 적용

lwj0831 committed:
fix(product): withId메서드 내 로직 오류 수정

lwj0831 committed:
refactor(exception): ErrorCode 리팩토링

- HttpStatus 필드 추가로 상태코드 정보 포함
- errorCode 필드로 서비스 자체 에러코드 정보 포함
- 의미있는 비즈니스 예외의 명확한 에러메세지 작성

lwj0831 committed:
feat(product): 정렬할 필드값에 따른 전략패턴 도입"

- 기존의 로직은 switch문에 따른 정렬 필드값 분기의 로직이라 복잡도가
  높았음
- 전략 패턴을 도입하여 기존 로직 개선 목적

lwj0831 committed:
refactor: PagedResult 정적 팩터리 메서드 수정

- 생성자로 대체 가능한지 고민해보았으나 this()전 로직 추가 불가능하여
  정적팩터리 메서드로 구현

lwj0831 committed:
chore: ErrorCode 필드명 message->errorMessage로 바뀐 부분 반영

lwj0831 committed:
refactor(product): ProductRepository 추상화 및 InMemoryProductRepository
리팩토링

- 추후 DB 연결 확장성을 고려한 추상화를 위해 ProductRepository
  인터페이스 정의
- InMemoryProductRepository 인터페이스에 맞게 Override 구현
- 메서드 parameter에 대해 Objects.requireNonNull()로 최소한의
  방어코드 구현

lwj0831 committed:
chore(product): NullPointerException 에러 메세지 영어에서 한글로 변경

lwj0831 committed:
chore(product): GetproductResDto 수정(Product 클래스에서 레코드로 변경한 사항 반영)

lwj0831 committed:
refactor(product): ProductService 리팩토링

- CreateProductReqDto, UpdateProdctReqDto에 imageUrl 필드 추가
- Product.of() 정적 팩터리 메서드 대신 DTO를 매개변수로 받는 새로운 정적
  팩터리 메서드 정의하면 좋을지 고민
- 이후 많은 DTO에 해당하는 팩터리 메서드로 인해 Product 클래스가
  무거워질 것이라 생각하여 기존의 of를 사용

lwj0831 committed:
refactor(product): ProductSortStrategyFactory메서드 getStrategy에 예외 발생 명확하게 표현

lwj0831 committed:
refactor(product): 에외 처리 리팩토링

- 기존의 HttpStatus로 대체 가능했던 ErrorCode 변경
- 서비스 내 자체 에러코드 필드 추가
- 비즈니스적으로 의미있는 예외에 대한 HttpStatus, errorCode,
  errorMessage 정의

lwj0831 committed:
refactor(product): ProductContorller 리팩토링

- createProduct시 생성된 객체의 URI 반환하도록 변경(그에 따른 Service,
  Repository도 생성된 Product객체의 Id를 반환하도록 변경)
- 조회, 변경, 삭제 API 개선

lwj0831 committed:
feat: Pagination sort필드 필터링을 위한 MethodArgumentResolver 도입"

- ModelAttribute와 고민하였으나 MethodArgumentResolver로 Pagination
  정렬용 어노테이션 정의하는게 재사용성 측면에서 좋다고 판단
- WebConfig로 SortParamMethodArgumentResolver 추가
- Custom Method Parameter 어노테이션인 SortParam 정의
- dto용 SortInfo 레코드 정의
- SortParamArgumentResolver 커스텀 MethodArugmentResolver 정의

lwj0831 committed:
refactor(product): Pagination 정렬 처리 부분 리팩토링

- Controller에서 커스텀 MethodArgumentResolver 사용하여 정렬 필드 처리하도록 변경
- 전략에 따른 Comparator 선정의 책임 Repository에서 Service로 이동
- ProductSortStrategyFactory내 Comparator 반환하는 정적 팩터리 메서드
  추가

lwj0831 committed:
feat(product): Product validation unit test 작성

lwj0831 committed:
refactor: Pagination 리팩터링

- Spring Data의 Pageable과 유사한 역할의 PageRequest 정의하여
  Pagination관련 파라미터 offset,size, sortInfo 하나의 DTO로 관리하도록 변경
- Controller, Service, Repository에서 관련 로직 변경
- 기존의 sortInfo 전략패턴에 따른 Comparator선정
  Service->Repository단으로 이동
- WebConfig에서 MethodArgumentResolver 이름 변경

lwj0831 committed:
chore(build): lombok 의존성 제거

lwj0831 committed:
refactor: BuisinessException 발생 시 디버깅 용 오류 로깅 출력 기능 추가

lwj0831 committed:
fix(product): 상품 수정 시 id값 null이 되는 bug fix

lwj0831 committed:
test(product): ProductService unit 테스트 코드 작성

- 정상 처리 케이스, 일반적인 오류 케이스 추가
- 동시성 처리 정상 수행 여부 확인할 수 있는 테스트 케이스 추가

lwj0831 committed:
chore(product): 상품 조회 Dto(GetProductResDto)에 누락된 imagUrl필드 추가

lwj0831 committed:
refactor: 관리자 페이지 api 수정 및 form html 수정

-imageUrl 항목 추가

lwj0831 committed:
docs: README.md에 상품 조회 시 Pagingation 지원 부분 기술

lwj0831 committed:
chore: 자잘한 부분 수정

- PageRequest의 pageSize 최소 1이상임을 검증하도록 변경
- GlobalExceptionHandler 로그 출력 시 문구 변경

lwj0831 committed:
chore: favicon.io 요청 처리 빈 값을 보낼 수 있게 추가

lwj0831 committed:
refactor: common을 global 패키지의 하위패키지로 리팩토링

도메인에 종속되지 않는 일반 클래스에 대한 패키지 잘게 분리 위함

lwj0831 committed:
Merge pull request #2 from lwj0831/refactor/product

Refactor/product

woowahan-justin commented on src/main/java/gift/global/exception/ErrorCode.java:
```diff
@@ -0,0 +1,32 @@
+
+import org.springframework.http.HttpStatus;
+
+public enum ErrorCode {
```
명확한 코드 메시지 활용 좋습니다

liquidjoo replied:
명확한 코드 메시지 활용 좋습니다

woowahan-justin commented on src/main/java/gift/product/domain/Product.java:
```diff
@@ -1,34 +1,76 @@
-        this.name = name;
-        this.price = price;
-        this.description = description;
+  public static Product withId(Long id, String name, Integer price, String description, String imageUrl){
+    validateId(id);
+    return new Product(id, name, price,description,imageUrl);
+  }
+  public static Product withId(Long id, Product product){
+    validateId(id);
+    return new Product(id, product.name, product.price, product.description, product.imageUrl);
+  }
+
```
정적 팩토리 메서드 패턴을 언제 사용하는 것이 좋을까요? 사용한다면 생성자는 어떻게 처리하는 것이 좋을까요?
그렇다면 생성자의 역할은 무엇일까요?

woowahan-justin commented on src/test/java/gift/product/domain/ProductTest.java:
```diff
@@ -0,0 +1,78 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class ProductTest {
```
꼼꼼한 테스트 케이스 좋습니다 👍

liquidjoo replied:
꼼꼼한 테스트 케이스 좋습니다 👍

woowahan-justin commented on src/test/java/gift/product/domain/ProductTest.java:
```diff
@@ -0,0 +1,78 @@
+    Product product = Product.of("applePie", 3000, "sweet honey applePie", "http://aaa.com/apple.png");
+
+    assertNull(product.id());
+    assertEquals("applePie", product.name());
+    assertEquals(3000, product.price());
+    assertEquals("sweet honey applePie", product.description());
+    assertEquals("http://aaa.com/apple.png", product.imageUrl());
```
assertAll 이라는 것도 좋재하는데요
한번 사용해보아도 좋을 것 같네요 :)

woowahan-justin commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -1,60 +1,66 @@
+  }
+
+  @GetMapping("/{productId}")
+  public ResponseEntity<GetProductResDto> getProductById(@PathVariable Long productId) {
```
스프링의 보안이 강화되면서 `@PathVariable` 과 같이 요청 값을 사용 시엔 어떤 이름인지 명확히 사용해주어야 합니다
e.g) `@PathVariable(name = "productId")` 이와 같이 사용하지 않을 때 실행 시 오류가 발생할 수 있기 때문에 처리해주는 것이 좋을 것 같아요.
요청 드렸던 페이지 접근 에러 이슈는 해당 이슈입니다 :)

woowahan-justin reviewed (approved):
안녕하세요 원준님 
2단계 피드백 잘해주셨어요
몇 가지 코멘트 남겼습니다 확인하고 다음 단계에 반영해보면 좋을 것 같아요
다음 단계 진행해 주세요 🙇

liquidjoo commented on src/main/java/gift/product/controller/ProductController.java:
```diff
@@ -1,60 +1,66 @@
+  }
+
+  @GetMapping("/{productId}")
+  public ResponseEntity<GetProductResDto> getProductById(@PathVariable Long productId) {
```
스프링의 보안이 강화되면서 `@PathVariable` 과 같이 요청 값을 사용 시엔 어떤 이름인지 명확히 사용해주어야 합니다
e.g) `@PathVariable(name = "productId")` 이와 같이 사용하지 않을 때 실행 시 오류가 발생할 수 있기 때문에 처리해주는 것이 좋을 것 같아요.
요청 드렸던 페이지 접근 에러 이슈는 해당 이슈입니다 :)

lwj0831 replied:
필드 검증 오류 발생 시 처리할 페이지가 없어서 나타나는 문제인줄 알았는데 PathVariable의 문제인지 피드백을 통해 알게 됐습니다! 
프론트엔드와 API연동 후 테스트할 때 PathVariable의 name을 생략해서 해당 문제 경험해본 적이 있는데 이번에는 생각을 못했습니다😅
피드백 덕분에 다시 한번 되새길 수 있었습니다!

liquidjoo commented on src/main/java/gift/product/domain/Product.java:
```diff
@@ -1,34 +1,76 @@
-        this.name = name;
-        this.price = price;
-        this.description = description;
+  public static Product withId(Long id, String name, Integer price, String description, String imageUrl){
+    validateId(id);
+    return new Product(id, name, price,description,imageUrl);
+  }
+  public static Product withId(Long id, Product product){
+    validateId(id);
+    return new Product(id, product.name, product.price, product.description, product.imageUrl);
+  }
+
+  public static Product of(String name, Integer price, String description, String imageUrl){
+    return new Product(null, name, price,description,imageUrl);
+  }
```
정적 팩토리 메서드 패턴을 언제 사용하는 것이 좋을까요? 사용한다면 생성자는 어떻게 처리하는 것이 좋을까요?
그렇다면 생성자의 역할은 무엇일까요?

lwj0831 replied:
정적 팩터리 메서드 패턴에 대한 질문 1단계 피드백 시에도 받았는데 답변으로 
정적 팩토리 메서드의 경우 생성자와 달리 메서드명을 정할 수 있어 메소드명을 통해 좀 더 명확한 메서드의 동작을 유추할 수 있고, 추가적인 복잡한 로직을 넣는 경우도 생성자에 비해 적합하다고 알고 있습니다.

도메인 내에서 필드 validation을 하는 로직의 경우에는 생성자보다는 정적 팩터리 메서드를 사용하는 것이 좋다고 판단하였고 리팩토링하였습니다.
그런데 복잡한 로직은 아니어서 생성자로 만들어도 괜찮을 것 같은데 정적 팩터리 메서드를 추가할 정도의 복잡한 로직의 기준을 어떻게 정해야할지 모르겠습니다..!

그래서 해당 코드의 경우 필드에 대한 검증 코드를 포함하는 이유에서 정적 팩터리 메서드를 선택하였는데 생성자로 구현하는게 더 적절했는지 궁금합니다..!

liquidjoo commented on src/test/java/gift/product/domain/ProductTest.java:
```diff
@@ -0,0 +1,78 @@
+    Product product = Product.of("applePie", 3000, "sweet honey applePie", "http://aaa.com/apple.png");
+
+    assertNull(product.id());
+    assertEquals("applePie", product.name());
+    assertEquals(3000, product.price());
+    assertEquals("sweet honey applePie", product.description());
+    assertEquals("http://aaa.com/apple.png", product.imageUrl());
```
assertAll 이라는 것도 좋재하는데요
한번 사용해보아도 좋을 것 같네요 :)

lwj0831 replied:
assertAll으로 그룹핑할 수 있는 부분들 리팩토링하였습니다

익숙한 메서드들로만 테스트 케이스 작성하고 있었는데 이번 기회에 assertAll외에도 도입하여 개선할 수 있는 부분있는지 확인해보겠습니다!

liquidjoo reviewed (approved):
안녕하세요 원준님
2단계 피드백 잘해주셨어요
몇 가지 코멘트 남겼습니다 확인하고 다음 단계에 반영해보면 좋을 것 같아요
다음 단계 진행해 주세요 🙇
